### ISSUE TITLE
WebGPU: RenderTarget is upside down

### ISSUE THREAD

**Issue by erikdubbelboer** [playcanvas/engine#6334]
The `flipY` of `pc.RenderTarget` is inverted for WebGPU.

Project that reproduces the bug: https://playcanvas.com/project/1176285/overview/webgpu-rendertarget-bug

WebGL:
<img width="1026" alt="Screenshot 2024-05-02 at 09 46 54" src="https://github.com/playcanvas/engine/assets/522870/db158425-474f-4369-b1bc-db7a46fc0dfb">

WebGPU:
<img width="1024" alt="Screenshot 2024-05-02 at 09 47 01" src="https://github.com/playcanvas/engine/assets/522870/0870c4b7-e7a1-4df3-b85e-aa5d8653d9ee">

Script used to generate the image:
```js
var Render = pc.createScript('render');

Render.prototype.initialize = function () {
    const device = this.app.graphicsDevice;

    this.colorBuffer = new pc.Texture(device, {
        width: 400,
        height: 400,
        format: pc.PIXELFORMAT_R8_G8_B8_A8,
        mipmaps: true,
        anisotropy: 1,
        minFilter: pc.FILTER_LINEAR_MIPMAP_NEAREST,
        magFilter: pc.FILTER_LINEAR,
        addressU: pc.ADDRESS_CLAMP_TO_EDGE,
        addressV: pc.ADDRESS_CLAMP_TO_EDGE,
        addressW: pc.ADDRESS_CLAMP_TO_EDGE,
    });

    const renderTarget = new pc.RenderTarget(device, this.colorBuffer, {
        depth: true,
        flipY: true, // Bug can be fixed using !device.isWebGPU
        samples: 1,
    });

    this.camera = this.app.root.findByName('Camera2').camera;

    this.camera.renderTarget = renderTarget;

    this.frameCounter = 0;
};

Render.prototype.update = function () {
    this.frameCounter++;

    // Wait a couple of frames to do this, doing it right away will fail.
    if (this.frameCounter == 10) {
        // This line isn't needed on WebGL, but on WebGPU causes:
        // Texture (unlabeled 400x400 px, TextureFormat::RGBA8Unorm)] usage (TextureUsage::(TextureBinding|RenderAttachment)) includes writable usage and another usage in the same synchronization scope.
        this.camera.renderTarget = null;

        this.entity.element.texture = this.colorBuffer;
    }
};
```

There is a simple fix to set `flipY` to `!device.isWebGPU`, but it's probably better if this gets fixed.

As you can see in the script another minor issue is that the `renderTarget` can't be in use by a camera at the same time as being used as a texture in WebGPU, with WebGL this is fine. I understand this is not something that can be fixed, but it's maybe nice to have documented somewhere if people search for this error.

**Comment by mvaligursky** [comment]
Yep, thanks, I'm aware of this, and have tried to fix it, but do not have a generic solution yet - there are many parts to it.

As you said, there is this workaround currently: `set flipY to !device.isWebGPU` but this depends on what gets rendered to the target. If forward renderer renders to it, the flip is internally handled by the adjusted projection matrix as well, but when custom shaders are used, this is different. Extra complication comes with cubemaps too.

**Comment by mvaligursky** [comment]
Closing as addressed by #9101 and #9105.

`RenderTarget#flipY` is relative to the native orientation of each graphics API, so the same value can result in different stored image orientations on WebGL2 and WebGPU. It is now deprecated in favor of the explicit, cross-API `origin` option.

For the render-to-texture case in this issue, where the result is displayed as a regular texture, use:

```js
const renderTarget = new pc.RenderTarget({
    colorBuffer: this.colorBuffer,
    depth: true,
    origin: pc.RENDERTARGET_ORIGIN_TOP,
    samples: 1
});
```

`RENDERTARGET_ORIGIN_TOP` stores the image in normal image orientation on both WebGL2 and WebGPU. Internally, this flips rendering on WebGL2 and does not flip it on WebGPU. `RENDERTARGET_ORIGIN_BOTTOM` supports consumers expecting WebGL-style bottom-origin storage, while the default `RENDERTARGET_ORIGIN_NATIVE` preserves API-native orientation and can therefore differ between backends.

The behavior is covered by RenderTarget origin unit tests and a cross-backend render-target orientation test example. It is available from `v2.22.0-beta.3`.

The WebGPU validation error mentioned separately is an expected render feedback hazard: a texture cannot be sampled while the same subresource is being used as a render attachment in that render pass. Detach it from the writing camera, exclude the consuming element from that camera layer, or use separate or ping-pong render targets. WebGL appearing to accept that feedback loop should not be relied upon.
